language/compiler/bytecode_source_map/src/mapping.rs

```diff
@@ -0,0 +1,486 @@
+            .struct_handle_at(struct_def.struct_handle)
+            .type_formals
+            .len();
+
+        if type_param_offset >= num_type_formals {
+            bail!("Type parameter offset out of bounds while trying to mark a struct's type parameters")
+        }
+
+        self.struct_marks
+            .entry(struct_definition_index.0)
+            .or_insert_with(StructMarking::new)
+            .type_param(type_param_offset, message);
+        Ok(())
+    }
+
+    //***************************************************************************
+    // Disassemblers
+    //***************************************************************************
+
+    pub fn disassemble_function_def(
+        &self,
+        function_definition_index: FunctionDefinitionIndex,
+    ) -> Result<String> {
+        let function_definition = self.bytecode.function_def_at(function_definition_index);
+        let function_handle = self
+            .bytecode
+            .function_handle_at(function_definition.function);
+        let function_signature = self
+            .bytecode
+            .function_signature_at(function_handle.signature);
+
+        let function_source_map = self
+            .source_map
+            .get_function_source_map(function_definition_index)?;
+
+        if self.options.only_public && !function_definition.is_public() {
+            return Ok("".to_string());
+        }
+
+        let visibility_modifier = if function_definition.is_native() {
+            "native "
+        } else if function_definition.is_public() {
+            "public "
+        } else {
+            ""
+        };
+
+        let ty_params: Vec<String> = function_source_map
+            .type_parameters
+            .iter()
+            .zip(function_signature.type_formals.iter())
+            .map(|((name, _), kind)| format!("{}: {:#?}", name.as_str(), kind))
+            .collect();
+        let name = self
+            .bytecode
+            .identifier_at(function_handle.name)
+            .to_string();
+        let ret_type: Vec<String> = function_signature
+            .return_types
+            .iter()
+            .cloned()
+            .map(|sig_token| {
+                let sig_tok_str =
+                    self.disassemble_sig_tok(sig_token, &function_source_map.type_parameters)?;
+                Ok(sig_tok_str.to_string())
+            })
+            .collect::<Result<Vec<String>>>()?;
+        let args: Vec<String> = function_signature
+            .arg_types
+            .iter()
+            .zip(function_source_map.locals.iter())
+            .map(|(typ, (name, _))| {
+                let ty_str =
+                    self.disassemble_sig_tok(typ.clone(), &function_source_map.type_parameters)?;
+                Ok(format!("{}: {}", name.to_string(), ty_str))
+            })
+            .collect::<Result<Vec<String>>>()?;
+        Ok(format!(
+            "{visibility_modifier}{name}<{ty_params}>({args}):{ret_type}",
```

Comment:
Would adding the `<>` around `ty_params` above only in the case where they're non-empty allow you to prevent the funky `<>` printing?

Reply:
yes! I was thinking of doing that along with not printing '{' and '}' for empty (i.e. native) structs.